File: generated_code/GPT4/Java/091_SplayTree.java
public class SplayTree {
    private class Node {
        int key;
        Node left, right, parent;

        Node(int key) {
            this.key = key;
        }
    }

    private Node root;

    // Right rotation
    private void rotateRight(Node x) {
        Node y = x.left;
        if (y == null) return;
        x.left = y.right;
        if (y.right != null)
            y.right.parent = x;
        y.parent = x.parent;
        if (x.parent == null)
            root = y;
        else if (x == x.parent.left)
            x.parent.left = y;
        else
            x.parent.right = y;
        y.right = x;
        x.parent = y;
    }

    // Left rotation
    private void rotateLeft(Node x) {
        Node y = x.right;
        if (y == null) return;
        x.right = y.left;
        if (y.left != null)
            y.left.parent = x;
        y.parent = x.parent;
        if (x.parent == null)
            root = y;
        else if (x == x.parent.left)
            x.parent.left = y;
        else
            x.parent.right = y;
        y.left = x;
        x.parent = y;
    }

    // Splay operation: brings x to the root
    private void splay(Node x) {
        if (x == null) return;
        while (x.parent != null) {
            if (x.parent.parent == null) {
                // Zig
                if (x.parent.left == x)
                    rotateRight(x.parent);
                else
                    rotateLeft(x.parent);
            } else if (x.parent.left == x && x.parent.parent.left == x.parent) {
                // Zig-Zig (left-left)
                rotateRight(x.parent.parent);
                rotateRight(x.parent);
            } else if (x.parent.right == x && x.parent.parent.right == x.parent) {
                // Zig-Zig (right-right)
                rotateLeft(x.parent.parent);
                rotateLeft(x.parent);
            } else if (x.parent.left == x && x.parent.parent.right == x.parent) {
                // Zig-Zag (left-right)
                rotateRight(x.parent);
                rotateLeft(x.parent);
            } else {
                // Zig-Zag (right-left)
                rotateLeft(x.parent);
                rotateRight(x.parent);
            }
        }
    }

    // Search for a key, splay the found node (or last accessed node)
    // Returns key if found, else null
    public Integer search(int key) {
        Node x = root;
        Node last = null;
        while (x != null) {
            last = x;
            if (key < x.key)
                x = x.left;
            else if (key > x.key)
                x = x.right;
            else
                break;
        }
        if (last != null)
            splay(last);
        if (x != null && x.key == key)
            return x.key;
        return null;
    }

    // Insert a key if not present
    public void insert(int key) {
        if (root == null) {
            root = new Node(key);
            return;
        }
        Node x = root, parent = null;
        while (x != null) {
            parent = x;
            if (key < x.key)
                x = x.left;
            else if (key > x.key)
                x = x.right;
            else {
                splay(x); // Key already exists, splay it
                return;
            }
        }
        Node newNode = new Node(key);
        newNode.parent = parent;
        if (key < parent.key)
            parent.left = newNode;
        else
            parent.right = newNode;
        splay(newNode);
    }

    // Delete a key if present
    public void delete(int key) {
        if (root == null) return;
        Integer found = search(key);
        if (found == null) return; // Not found

        // Now root holds the node to delete
        Node left = root.left;
        Node right = root.right;

        if (left != null) left.parent = null;
        if (right != null) right.parent = null;

        // Remove root
        root = null;

        if (left == null) {
            root = right;
        } else {
            // Find the max node in left subtree
            Node max = left;
            while (max.right != null) max = max.right;
            splay(max); // Splay max to root of left subtree
            // Now max has no right child
            max.right = right;
            if (right != null) right.parent = max;
            root = max;
        }
    }

    // Optional: For testing, print tree (inorder)
    private void inorder(Node x) {
        if (x != null) {
            inorder(x.left);
            System.out.print(x.key + " ");
            inorder(x.right);
        }
    }

    // For testing purposes
    public void printInOrder() {
        inorder(root);
        System.out.println();
    }

    // Main method for simple testing
    public static void main(String[] args) {
        SplayTree set = new SplayTree();
        set.insert(10);
        set.insert(20);
        set.insert(5);
        set.insert(8);
        set.insert(15);
        set.printInOrder(); // Should print 5 8 10 15 20

        System.out.println(set.search(15)); // 15, moved to root
        set.printInOrder();

        set.delete(10);
        set.printInOrder(); // 5 8 15 20

        System.out.println(set.search(99)); // null
    }
}

File: generated_code/GPT4/Python/091_SplayTree.py
class SplayTree:
    class Node:
        __slots__ = ('key', 'left', 'right', 'parent')

        def __init__(self, key, parent=None):
            self.key = key
            self.left = None
            self.right = None
            self.parent = parent

    def __init__(self):
        self.root = None

    def _rotate_left(self, x):
        y = x.right
        if y is None:
            return
        x.right = y.left
        if y.left:
            y.left.parent = x
        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        y.left = x
        x.parent = y

    def _rotate_right(self, x):
        y = x.left
        if y is None:
            return
        x.left = y.right
        if y.right:
            y.right.parent = x
        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
        y.right = x
        x.parent = y

    def _splay(self, x):
        while x.parent:
            p = x.parent
            gp = p.parent
            if not gp:
                # Zig
                if x == p.left:
                    self._rotate_right(p)
                else:
                    self._rotate_left(p)
            elif x == p.left and p == gp.left:
                # Zig-zig
                self._rotate_right(gp)
                self._rotate_right(p)
            elif x == p.right and p == gp.right:
                # Zig-zig
                self._rotate_left(gp)
                self._rotate_left(p)
            elif x == p.right and p == gp.left:
                # Zig-zag
                self._rotate_left(p)
                self._rotate_right(gp)
            else:
                # x == p.left and p == gp.right
                self._rotate_right(p)
                self._rotate_left(gp)

    def search(self, key):
        node = self.root
        last = None
        while node:
            last = node
            if key == node.key:
                self._splay(node)
                return True
            elif key < node.key:
                node = node.left
            else:
                node = node.right
        if last:
            self._splay(last)
        return False

    def insert(self, key):
        if not self.root:
            self.root = self.Node(key)
            return
        node = self.root
        parent = None
        while node:
            parent = node
            if key == node.key:
                self._splay(node)
                return
            elif key < node.key:
                node = node.left
            else:
                node = node.right
        new_node = self.Node(key, parent)
        if key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node
        self._splay(new_node)

    def _subtree_min(self, node):
        while node.left:
            node = node.left
        return node

    def delete(self, key):
        if not self.root:
            return
        node = self.root
        while node:
            if key == node.key:
                self._splay(node)
                break
            elif key < node.key:
                node = node.left
            else:
                node = node.right
        else:
            # Node not found; splay the last accessed node (or its parent)
            if node is None and self.root:
                # last = parent of where the key would be
                # Already at root
                return
            return
        # Node is at root
        if not node.left:
            self._replace(node, node.right)
        elif not node.right:
            self._replace(node, node.left)
        else:
            min_right = self._subtree_min(node.right)
            if min_right.parent != node:
                self._replace(min_right, min_right.right)
                min_right.right = node.right
                if min_right.right:
                    min_right.right.parent = min_right
            self._replace(node, min_right)
            min_right.left = node.left
            if min_right.left:
                min_right.left.parent = min_right
        del node

    def _replace(self, u, v):
        if not u.parent:
            self.root = v
        elif u == u.parent.left:
            u.parent.left = v
        else:
            u.parent.right = v
        if v:
            v.parent = u.parent

    # Optional: For debugging and testing
    def inorder(self):
        result = []
        def _inorder(node):
            if node:
                _inorder(node.left)
                result.append(node.key)
                _inorder(node.right)
        _inorder(self.root)
        return result

# Example Usage:
if __name__ == "__main__":
    st = SplayTree()
    st.insert(5)
    st.insert(3)
    st.insert(8)
    st.insert(7)
    st.insert(10)
    print("Inorder after inserts:", st.inorder())  # Should print [3, 5, 7, 8, 10]
    print("Search 7:", st.search(7))  # Should splay 7 to root
    print("Root after searching 7:", st.root.key)
    st.delete(5)
    print("Inorder after deleting 5:", st.inorder())  # Should print [3, 7, 8, 10]
    print("Search 100:", st.search(100))  # Not found, splay last accessed node
    print("Root after searching 100 (not found):", st.root.key)